src/ProtoBuf.FSharp/CodeGen.fs

```diff
@@ -0,0 +1,133 @@
+module private ProtoBuf.FSharp.CodeGen
+
+open System
+open FSharp.Reflection
+open System.Collections.Concurrent
+open System.Reflection
+open System.Reflection.Emit
+
+
+let private emitFieldAssignments (gen : ILGenerator, assigns : struct (FieldInfo * MethodInfo)[]) =
+    for (fi, getValue) in assigns do
+        if fi.IsStatic then
+            gen.EmitCall(OpCodes.Call, getValue, null)
+            gen.Emit(OpCodes.Stsfld, fi)
+        else
+            gen.Emit(OpCodes.Dup)
+            gen.EmitCall(OpCodes.Call, getValue, null)
+            gen.Emit(OpCodes.Stfld, fi)
+
+
+let private emitFactory (resultType : Type, assigns) =
+    let factoryMethod = DynamicMethod("factory_" + resultType.FullName, resultType, [| |], true)
+    let gen = factoryMethod.GetILGenerator()
+
+    match resultType.GetConstructor [| |] with
+    | null when FSharpType.IsRecord (resultType, true) ->
+        for pi in FSharpType.GetRecordFields (resultType, true) do
+            let tp = pi.PropertyType
+            match ZeroValues.calculateIfApplicable tp with
+            | Some getValue -> gen.EmitCall(OpCodes.Call, getValue, null)
+            | _ when tp.IsValueType ->
+                let cell = gen.DeclareLocal(tp)
+                gen.Emit(OpCodes.Ldloca_S, cell)
```

Comment:
That was just a short lazy way which works with all of { structs, primitive types, enums }. If I understand correctly, value types may in principle have parameterless constructor defined, but it's a very exotic thing. Although, flinging uninitialized chunks of memory around as structs may backfire in some spectacular way, so doing OpCodes.Initobj may be a very good idea. And it should be OpCodes.Ldloc instead of OpCodes.Ldloca. I'll fix it.